priv/post.ex.eex

```diff
@@ -66,4 +64,23 @@ defmodule <%= context.module_name %> do
   defp build_url(host, %{:proto => proto, :port => port}) do
     "#{proto}://#{host}:#{port}/"
   end
+
+  defp do_request(client, method, url, payload, headers, options) do
+    {mod, fun} = Map.fetch(client, :http_client)
+    apply(mod, fun, [method, url, payload, headers, options])
+  end
+
+  defp encode!(client, payload) do
+    {mod, fun} = client
+      |> Map.fetch(:encode)
+      |> Map.fetch(:<%= context.encode %>)
+    apply(mod, fun, [payload])
+  end
+
+  defp decode!(client, payload) do
+    {mod, fun} = client
+      |> Map.fetch(:decode)
+      |> Map.fetch(:<%= context.decode %>)
+    apply(mod, fun, [payload])
+  end
```

Comment:
I wonder if these functions should be moved to AWS.Client?

Eg.:
```elixir
AWS.Client.encode!(client, payload, :json)
AWS.Client.decode!(client, payload, :xml)
AWS.Client.request(client, method, url, payload, headers, options)
```

Reply:
I would say so, yes. In my mind, everything that is not auto-generated based on aws specs, should just delegate to aws-elixir.

I was even thinking that we should have called `AWS.JSON.encode!(client, payload)` but I like your version even more because your version assumes even less about aws-elixir. So to confirm:

Let's call `AWS.Client.encode!(client, payload, :json)` and friends and they will all be "straight-forward" implementations for now (i.e. just call poison, xmerl, etc) until we make it configurable on the aws-elixir side.

Btw, thanks for all of the work on this. There is some back and forth but this is not easy to figure out and I think the design will be much cleaner once we are done!

Reply:
That's great! People will be able to pass different implementations of AWS.Client. For instance, they could pass one that supports Instance Roles. Reaaaally nice!

Thank you for your help too! I am learning a lot! 😄 

I will push the changes soon!